File: Kernel Simulator/Network/Mail/MailDirectory/MailManager.vb

'    Kernel Simulator  Copyright (C) 2018-2021  EoflaOE
'
'    This file is part of Kernel Simulator
'
'    Kernel Simulator is free software: you can redistribute it and/or modify
'    it under the terms of the GNU General Public License as published by
'    the Free Software Foundation, either version 3 of the License, or
'    (at your option) any later version.
'
'    Kernel Simulator is distributed in the hope that it will be useful,
'    but WITHOUT ANY WARRANTY; without even the implied warranty of
'    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
'    GNU General Public License for more details.
'
'    You should have received a copy of the GNU General Public License
'    along with this program.  If not, see <https://www.gnu.org/licenses/>.

Imports System.Text
Imports MailKit
Imports MimeKit

Public Module MailManager

    Public ShowPreview As Boolean

    ''' <summary>
    ''' Lists messages
    ''' </summary>
    ''' <param name="PageNum">Page number</param>
    ''' <returns>String list of messages and preview (optional)</returns>
    ''' <exception cref="ArgumentException"></exception>
    Public Function MailListMessages(PageNum As Integer) As String
        If PageNum <= 0 Then PageNum = 1
        Wdbg("I", "Page number {0}", PageNum)
        If PageNum <= 0 Then
            Wdbg("E", "Trying to access page 0 or less than 0.")
            Throw New ArgumentException(DoTranslation("Page may not be negative or zero."))
            Return ""
        End If

        Dim EntryBuilder As New StringBuilder
        Dim MsgsLimitForPg As Integer = 10
        Dim FirstIndex As Integer = (MsgsLimitForPg * PageNum) - 10
        Dim LastIndex As Integer = (MsgsLimitForPg * PageNum) - 1
        Dim MaxMessagesIndex As Integer = IMAP_Messages.Count - 1
        Wdbg("I", "10 messages shown in each page. First message number in page {0} is {1} and last message number in page {0} is {2}", MsgsLimitForPg, FirstIndex, LastIndex)
        For i As Integer = FirstIndex To LastIndex
            If Not i > MaxMessagesIndex Then
                Wdbg("I", "Getting message {0}...", i)
                SyncLock IMAP_Client.SyncRoot
                    Dim Msg As MimeMessage
                    If Not IMAP_CurrentDirectory = "" And Not IMAP_CurrentDirectory = "Inbox" Then
                        Dim Dir As MailFolder = OpenFolder(IMAP_CurrentDirectory)
                        Msg = Dir.GetMessage(IMAP_Messages(i))
                    Else
                        Msg = IMAP_Client.Inbox.GetMessage(IMAP_Messages(i))
                    End If
                    Dim MsgFrom As String = Msg.From.ToString
                    Dim MsgSubject As String = Msg.Subject
                    Wdbg("I", "From {0}: {1}", MsgFrom, MsgSubject)
                    EntryBuilder.AppendLine($"- [{i + 1}] {Msg.From}: {Msg.Subject}")

                    'TODO: For more efficient preview, use the PREVIEW extension as documented in RFC-8970 (https://tools.ietf.org/html/rfc8970). However,
                    '      this is impossible at this time because no server and no client support this extension. It supports the LAZY modifier. It only
                    '      displays 200 character long body.
                    '      Concept: Msg.Preview(LazyMode:=True)
                    If ShowPreview Then
                        Dim MsgPreview As String = Msg.GetTextBody(Text.TextFormat.Text).Truncate(200)
                        If ColoredShell Then
                            EntryBuilder.AppendLine($"{New Color(ListValueColor).VTSequenceForeground}{MsgPreview}")
                            EntryBuilder.AppendLine($"{New Color(NeutralTextColor).VTSequenceForeground}")
                        Else
                            EntryBuilder.AppendLine(MsgPreview)
                        End If
                    End If
                End SyncLock
            Else
                Wdbg("W", "Reached max message limit. Message number {0}", i)
            End If
        Next
        Return EntryBuilder.ToString
    End Function

    ''' <summary>
    ''' Removes a message
    ''' </summary>
    ''' <param name="MsgNumber">Message number</param>
    ''' <returns>True if successful; False if unsuccessful</returns>
    ''' <exception cref="ArgumentException"></exception>
    ''' <exception cref="Exceptions.MailException"></exception>
    Public Function MailRemoveMessage(MsgNumber As Integer) As Boolean
        Dim Message As Integer = MsgNumber - 1
        Dim MaxMessagesIndex As Integer = IMAP_Messages.Count - 1
        Wdbg("I", "Message number {0}", Message)
        If Message < 0 Then
            Wdbg("E", "Trying to remove message 0 or less than 0.")
            Throw New ArgumentException(DoTranslation("Message number may not be negative or zero."))
            Return False
        ElseIf Message > MaxMessagesIndex Then
            Wdbg("E", "Message {0} not in list. It was larger than MaxMessagesIndex ({1})", Message, MaxMessagesIndex)
            Throw New Exceptions.MailException(DoTranslation("Message specified is not found."))
            Return False
        End If

        SyncLock IMAP_Client.SyncRoot
            If Not IMAP_CurrentDirectory = "" And Not IMAP_CurrentDirectory = "Inbox" Then
                'Remove message
                Dim Dir As MailFolder = OpenFolder(IMAP_CurrentDirectory)
                Wdbg("I", "Opened {0}. Removing {1}...", IMAP_CurrentDirectory, MsgNumber)
                Dir.AddFlags(IMAP_Messages(Message), MessageFlags.Deleted, True)
                Wdbg("I", "Removed.")
                Dir.Expunge()
            Else
                'Remove message
                IMAP_Client.Inbox.Open(FolderAccess.ReadWrite)
                Wdbg("I", "Removing {0}...", MsgNumber)
                IMAP_Client.Inbox.AddFlags(IMAP_Messages(Message), MessageFlags.Deleted, True)
                Wdbg("I", "Removed.")
                IMAP_Client.Inbox.Expunge()
            End If
        End SyncLock
        Return True
    End Function

    ''' <summary>
    ''' Removes all mail that the specified sender has sent
    ''' </summary>
    ''' <param name="Sender">The sender name</param>
    ''' <returns>True if successful; False if unsuccessful</returns>
    Public Function MailRemoveAllBySender(Sender As String) As Boolean
        Wdbg("I", "All mail by {0} will be removed.", Sender)
        Dim DeletedMsgNumber As Integer = 1
        Dim SteppedMsgNumber As Integer = 0
        For i As Integer = 0 To IMAP_Messages.Count
            Try
                SyncLock IMAP_Client.SyncRoot
                    Dim MessageId As UniqueId = IMAP_Messages(i)
                    Dim Msg As MimeMessage
                    If Not IMAP_CurrentDirectory = "" And Not IMAP_CurrentDirectory = "Inbox" Then
                        Dim Dir As MailFolder = OpenFolder(IMAP_CurrentDirectory)
                        Msg = Dir.GetMessage(MessageId)
                    Else
                        Msg = IMAP_Client.Inbox.GetMessage(MessageId)
                    End If
                    SteppedMsgNumber += 1

                    For Each address In Msg.From
                        If address.Name = Sender Then
                            If Not IMAP_CurrentDirectory = "" And Not IMAP_CurrentDirectory = "Inbox" Then
                                Dim Dir As MailFolder = OpenFolder(IMAP_CurrentDirectory)

                                'Remove message
                                Wdbg("I", "Opened {0}. Removing {1}...", IMAP_CurrentDirectory, Sender)
                                Dir.AddFlags(MessageId, MessageFlags.Deleted, True)
                                Wdbg("I", "Removed.")
                                Dir.Expunge()
                                Wdbg("I", "Message {0} from {1} deleted from {2}. {3} messages remaining to parse.", DeletedMsgNumber, Sender, IMAP_CurrentDirectory, IMAP_Messages.Count - SteppedMsgNumber)
                                Write(DoTranslation("Message {0} from {1} deleted from {2}. {3} messages remaining to parse."), True, ColTypes.Neutral, DeletedMsgNumber, Sender, IMAP_CurrentDirectory, IMAP_Messages.Count - SteppedMsgNumber)
                            Else
                                'Remove message
                                IMAP_Client.Inbox.Open(FolderAccess.ReadWrite)
                                Wdbg("I", "Removing {0}...", Sender)
                                IMAP_Client.Inbox.AddFlags(MessageId, MessageFlags.Deleted, True)
                                Wdbg("I", "Removed.")
                                IMAP_Client.Inbox.Expunge()
                                Wdbg("I", "Message {0} from {1} deleted from inbox. {2} messages remaining to parse.", DeletedMsgNumber, Sender, IMAP_Messages.Count - SteppedMsgNumber)
                                Write(DoTranslation("Message {0} from {1} deleted from inbox. {2} messages remaining to parse."), True, ColTypes.Neutral, DeletedMsgNumber, Sender, IMAP_Messages.Count - SteppedMsgNumber)
                            End If
                            DeletedMsgNumber += 1
                        End If
                    Next
                End SyncLock
            Catch ex As Exception
                WStkTrc(ex)
                Return False
            End Try
        Next
        Return True
    End Function

    ''' <summary>
    ''' Moves a message
    ''' </summary>
    ''' <param name="MsgNumber">Message number</param>
    ''' <param name="TargetFolder">Target folder</param>
    ''' <returns>True if successful; False if unsuccessful</returns>
    ''' <exception cref="ArgumentException"></exception>
    ''' <exception cref="Exceptions.MailException"></exception>
    Public Function MailMoveMessage(MsgNumber As Integer, TargetFolder As String) As Boolean
        Dim Message As Integer = MsgNumber - 1
        Dim MaxMessagesIndex As Integer = IMAP_Messages.Count - 1
        Wdbg("I", "Message number {0}", Message)
        If Message < 0 Then
            Wdbg("E", "Trying to move message 0 or less than 0.")
            Throw New ArgumentException(DoTranslation("Message number may not be negative or zero."))
            Return False
        ElseIf Message > MaxMessagesIndex Then
            Wdbg("E", "Message {0} not in list. It was larger than MaxMessagesIndex ({1})", Message, MaxMessagesIndex)
            Throw New Exceptions.MailException(DoTranslation("Message specified is not found."))
            Return False
        End If

        SyncLock IMAP_Client.SyncRoot
            If Not IMAP_CurrentDirectory = "" And Not IMAP_CurrentDirectory = "Inbox" Then
                'Move message
                Dim Dir As MailFolder = OpenFolder(IMAP_CurrentDirectory)
                Dim TargetF As MailFolder = OpenFolder(TargetFolder)
                Wdbg("I", "Opened {0}. Moving {1}...", IMAP_CurrentDirectory, MsgNumber)
                Dir.MoveTo(IMAP_Messages(Message), TargetF)
                Wdbg("I", "Moved.")
            Else
                'Move message
                Dim TargetF As MailFolder = OpenFolder(TargetFolder)
                Wdbg("I", "Moving {0}...", MsgNumber)
                IMAP_Client.Inbox.Open(FolderAccess.ReadWrite)
                IMAP_Client.Inbox.MoveTo(IMAP_Messages(Message), TargetF)
                Wdbg("I", "Moved.")
            End If
        End SyncLock
        Return True
    End Function

    ''' <summary>
    ''' Moves all mail that the specified sender has sent
    ''' </summary>
    ''' <param name="Sender">The sender name</param>
    ''' <param name="TargetFolder">Target folder</param>
    ''' <returns>True if successful; False if unsuccessful</returns>
    Public Function MailMoveAllBySender(Sender As String, TargetFolder As String) As Boolean
        Wdbg("I", "All mail by {0} will be moved.", Sender)
        Dim DeletedMsgNumber As Integer = 1
        Dim SteppedMsgNumber As Integer = 0
        For i As Integer = 0 To IMAP_Messages.Count
            Try
                SyncLock IMAP_Client.SyncRoot
                    Dim MessageId As UniqueId = IMAP_Messages(i)
                    Dim Msg As MimeMessage
                    If Not IMAP_CurrentDirectory = "" And Not IMAP_CurrentDirectory = "Inbox" Then
                        Dim Dir As MailFolder = OpenFolder(IMAP_CurrentDirectory)
                        Msg = Dir.GetMessage(MessageId)
                    Else
                        Msg = IMAP_Client.Inbox.GetMessage(MessageId)
                    End If
                    SteppedMsgNumber += 1

                    For Each address In Msg.From
                        If address.Name = Sender Then
                            If Not IMAP_CurrentDirectory = "" And Not IMAP_CurrentDirectory = "Inbox" Then
                                Dim Dir As MailFolder = OpenFolder(IMAP_CurrentDirectory)
                                Dim TargetF As MailFolder = OpenFolder(TargetFolder)
                                'Remove message
                                Wdbg("I", "Opened {0}. Moving {1}...", IMAP_CurrentDirectory, Sender)
                                Dir.MoveTo(MessageId, TargetF)
                                Wdbg("I", "Moved.")
                                Wdbg("I", "Message {0} from {1} moved from {2}. {3} messages remaining to parse.", DeletedMsgNumber, Sender, IMAP_CurrentDirectory, IMAP_Messages.Count - SteppedMsgNumber)
                                Write(DoTranslation("Message {0} from {1} moved from {2}. {3} messages remaining to parse."), True, ColTypes.Neutral, DeletedMsgNumber, Sender, IMAP_CurrentDirectory, IMAP_Messages.Count - SteppedMsgNumber)
                            Else
                                'Remove message
                                Dim TargetF As MailFolder = OpenFolder(TargetFolder)
                                Wdbg("I", "Moving {0}...", Sender)
                                IMAP_Client.Inbox.Open(FolderAccess.ReadWrite)
                                IMAP_Client.Inbox.MoveTo(MessageId, TargetF)
                                Wdbg("I", "Moved.")
                                Wdbg("I", "Message {0} from {1} moved. {2} messages remaining to parse.", DeletedMsgNumber, Sender, IMAP_Messages.Count - SteppedMsgNumber)
                                Write(DoTranslation("Message {0} from {1} moved. {2} messages remaining to parse."), True, ColTypes.Neutral, DeletedMsgNumber, Sender, IMAP_Messages.Count - SteppedMsgNumber)
                            End If
                            DeletedMsgNumber += 1
                        End If
                    Next
                End SyncLock
            Catch ex As Exception
                WStkTrc(ex)
                Return False
            End Try
        Next
        Return True
    End Function

End Module

File: public/Nitrocid.Addons/Nitrocid.Extras.MailShell/Tools/Directory/MailManager.cs
//
// Nitrocid KS  Copyright (C) 2018-2024  Aptivi
//
// This file is part of Nitrocid KS
//
// Nitrocid KS is free software: you can redistribute it and/or modify
// it under the terms of the GNU General Public License as published by
// the Free Software Foundation, either version 3 of the License, or
// (at your option) any later version.
//
// Nitrocid KS is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY, without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU General Public License for more details.
//
// You should have received a copy of the GNU General Public License
// along with this program.  If not, see <https://www.gnu.org/licenses/>.
//

using System;
using System.Data;
using System.Linq;
using KS.ConsoleBase.Colors;
using KS.ConsoleBase.Writers.ConsoleWriters;
using KS.Kernel.Debugging;
using KS.Kernel.Exceptions;
using KS.Languages;
using KS.Misc.Text;
using MailKit;
using MailKit.Net.Imap;
using MimeKit;
using Nitrocid.Extras.MailShell.Mail;

namespace Nitrocid.Extras.MailShell.Tools.Directory
{
    /// <summary>
    /// Mail management module
    /// </summary>
    public static class MailManager
    {

        /// <summary>
        /// Whether to show the message preview or not
        /// </summary>
        public static bool ShowPreview =>
            MailShellInit.MailConfig.ShowPreview;

        /// <summary>
        /// Lists messages
        /// </summary>
        /// <param name="PageNum">Page number</param>
        /// <exception cref="ArgumentException"></exception>
        public static void MailListMessages(int PageNum) => MailListMessages(PageNum, MailShellCommon.MaxMessagesInPage);

        /// <summary>
        /// Lists messages
        /// </summary>
        /// <param name="PageNum">Page number</param>
        /// <param name="MessagesInPage">Max messages in one page</param>
        /// <exception cref="ArgumentException"></exception>
        public static void MailListMessages(int PageNum, int MessagesInPage)
        {
            // Sanity checks for the page number
            if (PageNum <= 0)
                PageNum = 1;
            DebugWriter.WriteDebug(DebugLevel.I, "Page number {0}", PageNum);

            int MsgsLimitForPg = MessagesInPage;
            int FirstIndex = MsgsLimitForPg * PageNum - 10;
            int LastIndex = MsgsLimitForPg * PageNum - 1;
            int MaxMessagesIndex = MailShellCommon.IMAP_Messages.Count() - 1;
            DebugWriter.WriteDebug(DebugLevel.I, "10 messages shown in each page. First message number in page {0} is {1} and last message number in page {0} is {2}", MsgsLimitForPg, FirstIndex, LastIndex);
            for (int i = FirstIndex; i <= LastIndex; i++)
            {
                if (i <= MaxMessagesIndex)
                {
                    string MsgFrom = "";
                    string MsgSubject = "";
                    string MsgPreview = "";

                    // Getting information about the message is vital to display them.
                    DebugWriter.WriteDebug(DebugLevel.I, "Getting message {0}...", i);
                    lock (((ImapClient)((object[])MailShellCommon.Client.ConnectionInstance)[0]).SyncRoot)
                    {
                        MimeMessage Msg;
                        if (!string.IsNullOrEmpty(MailShellCommon.IMAP_CurrentDirectory) & !(MailShellCommon.IMAP_CurrentDirectory == "Inbox"))
                        {
                            var Dir = MailDirectory.OpenFolder(MailShellCommon.IMAP_CurrentDirectory);
                            Msg = Dir.GetMessage(MailShellCommon.IMAP_Messages.ElementAtOrDefault(i), default, MailShellCommon.Progress);
                        }
                        else
                        {
                            Msg = ((ImapClient)((object[])MailShellCommon.Client.ConnectionInstance)[0]).Inbox.GetMessage(MailShellCommon.IMAP_Messages.ElementAtOrDefault(i), default, MailShellCommon.Progress);
                        }
                        MsgFrom = Msg.From.ToString();
                        MsgSubject = Msg.Subject;
                        MsgPreview = Msg.GetTextBody(MimeKit.Text.TextFormat.Text).Truncate(200);
                    }
                    DebugWriter.WriteDebug(DebugLevel.I, "From {0}: {1}", MsgFrom, MsgSubject);

                    // Display them now.
                    TextWriterColor.WriteKernelColor($"- [{i + 1}/{MaxMessagesIndex + 1}] {MsgFrom}: ", false, KernelColorType.ListEntry);
                    TextWriterColor.WriteKernelColor(MsgSubject, true, KernelColorType.ListValue);
                    if (ShowPreview & !string.IsNullOrWhiteSpace(MsgPreview))
                    {
                        // For more efficient preview, use the PREVIEW extension as documented in RFC-8970 (https://tools.ietf.org/html/rfc8970). However,
                        // this is impossible at this time because no server and no client support this extension. It supports the LAZY modifier. It only
                        // displays 200 character long body.
                        //
                        // Concept: Msg.Preview(LazyMode:=True)
                        TextWriterColor.WriteKernelColor(MsgPreview, true, KernelColorType.ListValue);
                    }
                }
                else
                {
                    DebugWriter.WriteDebug(DebugLevel.W, "Reached max message limit. Message number {0}", i);
                }
            }
        }

        /// <summary>
        /// Removes a message
        /// </summary>
        /// <param name="MsgNumber">Message number</param>
        /// <returns>True if successful; False if unsuccessful</returns>
        /// <exception cref="ArgumentException"></exception>
        public static bool MailRemoveMessage(int MsgNumber)
        {
            int Message = MsgNumber - 1;
            int MaxMessagesIndex = MailShellCommon.IMAP_Messages.Count() - 1;
            DebugWriter.WriteDebug(DebugLevel.I, "Message number {0}", Message);
            if (Message < 0)
            {
                DebugWriter.WriteDebug(DebugLevel.E, "Trying to remove message 0 or less than 0.");
                throw new KernelException(KernelExceptionType.Mail, Translate.DoTranslation("Message number may not be negative or zero."));
            }
            else if (Message > MaxMessagesIndex)
            {
                DebugWriter.WriteDebug(DebugLevel.E, "Message {0} not in list. It was larger than MaxMessagesIndex ({1})", Message, MaxMessagesIndex);
                throw new KernelException(KernelExceptionType.Mail, Translate.DoTranslation("Message specified is not found."));
            }

            lock (((ImapClient)((object[])MailShellCommon.Client.ConnectionInstance)[0]).SyncRoot)
            {
                if (!string.IsNullOrEmpty(MailShellCommon.IMAP_CurrentDirectory) & !(MailShellCommon.IMAP_CurrentDirectory == "Inbox"))
                {
                    // Remove message
                    var Dir = MailDirectory.OpenFolder(MailShellCommon.IMAP_CurrentDirectory);
                    DebugWriter.WriteDebug(DebugLevel.I, "Opened {0}. Removing {1}...", MailShellCommon.IMAP_CurrentDirectory, MsgNumber);
                    Dir.Store(MailShellCommon.IMAP_Messages.ElementAtOrDefault(Message), new StoreFlagsRequest(StoreAction.Add, MessageFlags.Deleted));
                    DebugWriter.WriteDebug(DebugLevel.I, "Removed.");
                    Dir.Expunge();
                }
                else
                {
                    // Remove message
                    ((ImapClient)((object[])MailShellCommon.Client.ConnectionInstance)[0]).Inbox.Open(FolderAccess.ReadWrite);
                    DebugWriter.WriteDebug(DebugLevel.I, "Removing {0}...", MsgNumber);
                    ((ImapClient)((object[])MailShellCommon.Client.ConnectionInstance)[0]).Inbox.Store(MailShellCommon.IMAP_Messages.ElementAtOrDefault(Message), new StoreFlagsRequest(StoreAction.Add, MessageFlags.Deleted));
                    DebugWriter.WriteDebug(DebugLevel.I, "Removed.");
                    ((ImapClient)((object[])MailShellCommon.Client.ConnectionInstance)[0]).Inbox.Expunge();
                }
            }
            return true;
        }

        /// <summary>
        /// Removes all mail that the specified sender has sent
        /// </summary>
        /// <param name="Sender">The sender name</param>
        /// <returns>True if successful; False if unsuccessful</returns>
        public static bool MailRemoveAllBySender(string Sender)
        {
            DebugWriter.WriteDebug(DebugLevel.I, "All mail by {0} will be removed.", Sender);
            int DeletedMsgNumber = 1;
            int SteppedMsgNumber = 0;
            for (int i = 0; i <= MailShellCommon.IMAP_Messages.Count(); i++)
            {
                try
                {
                    lock (((ImapClient)((object[])MailShellCommon.Client.ConnectionInstance)[0]).SyncRoot)
                    {
                        var MessageId = MailShellCommon.IMAP_Messages.ElementAtOrDefault(i);
                        MimeMessage Msg;
                        if (!string.IsNullOrEmpty(MailShellCommon.IMAP_CurrentDirectory) & !(MailShellCommon.IMAP_CurrentDirectory == "Inbox"))
                        {
                            var Dir = MailDirectory.OpenFolder(MailShellCommon.IMAP_CurrentDirectory);
                            Msg = Dir.GetMessage(MessageId, default, MailShellCommon.Progress);
                        }
                        else
                        {
                            Msg = ((ImapClient)((object[])MailShellCommon.Client.ConnectionInstance)[0]).Inbox.GetMessage(MessageId, default, MailShellCommon.Progress);
                        }
                        SteppedMsgNumber += 1;

                        foreach (var address in Msg.From)
                        {
                            if (address.Name == Sender)
                            {
                                if (!string.IsNullOrEmpty(MailShellCommon.IMAP_CurrentDirectory) & !(MailShellCommon.IMAP_CurrentDirectory == "Inbox"))
                                {
                                    var Dir = MailDirectory.OpenFolder(MailShellCommon.IMAP_CurrentDirectory);

                                    // Remove message
                                    DebugWriter.WriteDebug(DebugLevel.I, "Opened {0}. Removing {1}...", MailShellCommon.IMAP_CurrentDirectory, Sender);
                                    Dir.Store(MessageId, new StoreFlagsRequest(StoreAction.Add, MessageFlags.Deleted));
                                    DebugWriter.WriteDebug(DebugLevel.I, "Removed.");
                                    Dir.Expunge();
                                    DebugWriter.WriteDebug(DebugLevel.I, "Message {0} from {1} deleted from {2}. {3} messages remaining to parse.", DeletedMsgNumber, Sender, MailShellCommon.IMAP_CurrentDirectory, MailShellCommon.IMAP_Messages.Count() - SteppedMsgNumber);
                                    TextWriterColor.Write(Translate.DoTranslation("Message {0} from {1} deleted from {2}. {3} messages remaining to parse."), DeletedMsgNumber, Sender, MailShellCommon.IMAP_CurrentDirectory, MailShellCommon.IMAP_Messages.Count() - SteppedMsgNumber);
                                }
                                else
                                {
                                    // Remove message
                                    ((ImapClient)((object[])MailShellCommon.Client.ConnectionInstance)[0]).Inbox.Open(FolderAccess.ReadWrite);
                                    DebugWriter.WriteDebug(DebugLevel.I, "Removing {0}...", Sender);
                                    ((ImapClient)((object[])MailShellCommon.Client.ConnectionInstance)[0]).Inbox.Store(MessageId, new StoreFlagsRequest(StoreAction.Add, MessageFlags.Deleted));
                                    DebugWriter.WriteDebug(DebugLevel.I, "Removed.");
                                    ((ImapClient)((object[])MailShellCommon.Client.ConnectionInstance)[0]).Inbox.Expunge();
                                    DebugWriter.WriteDebug(DebugLevel.I, "Message {0} from {1} deleted from inbox. {2} messages remaining to parse.", DeletedMsgNumber, Sender, MailShellCommon.IMAP_Messages.Count() - SteppedMsgNumber);
                                    TextWriterColor.Write(Translate.DoTranslation("Message {0} from {1} deleted from inbox. {2} messages remaining to parse."), DeletedMsgNumber, Sender, MailShellCommon.IMAP_Messages.Count() - SteppedMsgNumber);
                                }
                                DeletedMsgNumber += 1;
                            }
                        }
                    }
                }
                catch (Exception ex)
                {
                    DebugWriter.WriteDebugStackTrace(ex);
                    return false;
                }
            }
            return true;
        }

        /// <summary>
        /// Moves a message
        /// </summary>
        /// <param name="MsgNumber">Message number</param>
        /// <param name="TargetFolder">Target folder</param>
        /// <returns>True if successful; False if unsuccessful</returns>
        /// <exception cref="ArgumentException"></exception>
        public static bool MailMoveMessage(int MsgNumber, string TargetFolder)
        {
            int Message = MsgNumber - 1;
            int MaxMessagesIndex = MailShellCommon.IMAP_Messages.Count() - 1;
            DebugWriter.WriteDebug(DebugLevel.I, "Message number {0}", Message);
            if (Message < 0)
            {
                DebugWriter.WriteDebug(DebugLevel.E, "Trying to move message 0 or less than 0.");
                throw new KernelException(KernelExceptionType.Mail, Translate.DoTranslation("Message number may not be negative or zero."));
            }
            else if (Message > MaxMessagesIndex)
            {
                DebugWriter.WriteDebug(DebugLevel.E, "Message {0} not in list. It was larger than MaxMessagesIndex ({1})", Message, MaxMessagesIndex);
                throw new KernelException(KernelExceptionType.Mail, Translate.DoTranslation("Message specified is not found."));
            }

            lock (((ImapClient)((object[])MailShellCommon.Client.ConnectionInstance)[0]).SyncRoot)
            {
                if (!string.IsNullOrEmpty(MailShellCommon.IMAP_CurrentDirectory) & !(MailShellCommon.IMAP_CurrentDirectory == "Inbox"))
                {
                    // Move message
                    var Dir = MailDirectory.OpenFolder(MailShellCommon.IMAP_CurrentDirectory);
                    var TargetF = MailDirectory.OpenFolder(TargetFolder);
                    DebugWriter.WriteDebug(DebugLevel.I, "Opened {0}. Moving {1}...", MailShellCommon.IMAP_CurrentDirectory, MsgNumber);
                    Dir.MoveTo(MailShellCommon.IMAP_Messages.ElementAtOrDefault(Message), TargetF);
                    DebugWriter.WriteDebug(DebugLevel.I, "Moved.");
                }
                else
                {
                    // Move message
                    var TargetF = MailDirectory.OpenFolder(TargetFolder);
                    DebugWriter.WriteDebug(DebugLevel.I, "Moving {0}...", MsgNumber);
                    ((ImapClient)((object[])MailShellCommon.Client.ConnectionInstance)[0]).Inbox.Open(FolderAccess.ReadWrite);
                    ((ImapClient)((object[])MailShellCommon.Client.ConnectionInstance)[0]).Inbox.MoveTo(MailShellCommon.IMAP_Messages.ElementAtOrDefault(Message), TargetF);
                    DebugWriter.WriteDebug(DebugLevel.I, "Moved.");
                }
            }
            return true;
        }

        /// <summary>
        /// Moves all mail that the specified sender has sent
        /// </summary>
        /// <param name="Sender">The sender name</param>
        /// <param name="TargetFolder">Target folder</param>
        /// <returns>True if successful; False if unsuccessful</returns>
        public static bool MailMoveAllBySender(string Sender, string TargetFolder)
        {
            DebugWriter.WriteDebug(DebugLevel.I, "All mail by {0} will be moved.", Sender);
            int DeletedMsgNumber = 1;
            int SteppedMsgNumber = 0;
            for (int i = 0; i <= MailShellCommon.IMAP_Messages.Count(); i++)
            {
                try
                {
                    lock (((ImapClient)((object[])MailShellCommon.Client.ConnectionInstance)[0]).SyncRoot)
                    {
                        var MessageId = MailShellCommon.IMAP_Messages.ElementAtOrDefault(i);
                        MimeMessage Msg;
                        if (!string.IsNullOrEmpty(MailShellCommon.IMAP_CurrentDirectory) & !(MailShellCommon.IMAP_CurrentDirectory == "Inbox"))
                        {
                            var Dir = MailDirectory.OpenFolder(MailShellCommon.IMAP_CurrentDirectory);
                            Msg = Dir.GetMessage(MessageId, default, MailShellCommon.Progress);
                        }
                        else
                        {
                            Msg = ((ImapClient)((object[])MailShellCommon.Client.ConnectionInstance)[0]).Inbox.GetMessage(MessageId, default, MailShellCommon.Progress);
                        }
                        SteppedMsgNumber += 1;

                        foreach (var address in Msg.From)
                        {
                            if (address.Name == Sender)
                            {
                                if (!string.IsNullOrEmpty(MailShellCommon.IMAP_CurrentDirectory) & !(MailShellCommon.IMAP_CurrentDirectory == "Inbox"))
                                {
                                    var Dir = MailDirectory.OpenFolder(MailShellCommon.IMAP_CurrentDirectory);
                                    var TargetF = MailDirectory.OpenFolder(TargetFolder);
                                    // Remove message
                                    DebugWriter.WriteDebug(DebugLevel.I, "Opened {0}. Moving {1}...", MailShellCommon.IMAP_CurrentDirectory, Sender);
                                    Dir.MoveTo(MessageId, TargetF);
                                    DebugWriter.WriteDebug(DebugLevel.I, "Moved.");
                                    DebugWriter.WriteDebug(DebugLevel.I, "Message {0} from {1} moved from {2}. {3} messages remaining to parse.", DeletedMsgNumber, Sender, MailShellCommon.IMAP_CurrentDirectory, MailShellCommon.IMAP_Messages.Count() - SteppedMsgNumber);
                                    TextWriterColor.Write(Translate.DoTranslation("Message {0} from {1} moved from {2}. {3} messages remaining to parse."), DeletedMsgNumber, Sender, MailShellCommon.IMAP_CurrentDirectory, MailShellCommon.IMAP_Messages.Count() - SteppedMsgNumber);
                                }
                                else
                                {
                                    // Remove message
                                    var TargetF = MailDirectory.OpenFolder(TargetFolder);
                                    DebugWriter.WriteDebug(DebugLevel.I, "Moving {0}...", Sender);
                                    ((ImapClient)((object[])MailShellCommon.Client.ConnectionInstance)[0]).Inbox.Open(FolderAccess.ReadWrite);
                                    ((ImapClient)((object[])MailShellCommon.Client.ConnectionInstance)[0]).Inbox.MoveTo(MessageId, TargetF);
                                    DebugWriter.WriteDebug(DebugLevel.I, "Moved.");
                                    DebugWriter.WriteDebug(DebugLevel.I, "Message {0} from {1} moved. {2} messages remaining to parse.", DeletedMsgNumber, Sender, MailShellCommon.IMAP_Messages.Count() - SteppedMsgNumber);
                                    TextWriterColor.Write(Translate.DoTranslation("Message {0} from {1} moved. {2} messages remaining to parse."), DeletedMsgNumber, Sender, MailShellCommon.IMAP_Messages.Count() - SteppedMsgNumber);
                                }
                                DeletedMsgNumber += 1;
                            }
                        }
                    }
                }
                catch (Exception ex)
                {
                    DebugWriter.WriteDebugStackTrace(ex);
                    return false;
                }
            }
            return true;
        }

    }
}
